### PR TITLE
Include PackageMetadata in LF interface

### PR DESCRIPTION
**Pull request by cocreature**
I want to use this in codegens but more generally, I think it also
fits well into the scope of the iface library in that it’s only
metadata and relatively stable.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/Interface.scala [hunk 1]
```diff
@@ -42,12 +42,22 @@ object InterfaceType {
   }
 }
 
+// Duplicate of the one in com.daml.lf.language to separate Ast and Iface
```
Comment (cocreature) [inline daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/Interface.scala:45]: If someone tells me that it’s fine to just expose parts of the `Ast`, I’m also perfectly happy with that. I don’t know what our policy here is.

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/EnvironmentInterface.scala [hunk 2]
```diff
@@ -5,26 +5,34 @@ package com.daml.lf
...
-      typeDecls mapKeys (Identifier(packageId, _))
-    }.toMap)
+    (i +: o).foldLeft(EnvironmentInterface(Map.empty, Map.empty)) {
+      case (acc, Interface(packageId, optMetadata, typeDecls)) =>
+        acc.copy(
+          metadata =
+            optMetadata.fold(acc.metadata)(metadata => acc.metadata.updated(packageId, metadata)),
+          typeDecls = acc.typeDecls ++ typeDecls.mapKeys(Identifier(packageId, _)),
+        )
+    }
```
Comment (S11001001) [inline daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/EnvironmentInterface.scala:29]: I'd rather iterate over `i +: o` twice.

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/Interface.scala (hunk 1 above)
Comment (S11001001) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/Interface.scala:45]: I think this is fine. Maybe we'll add a type parameter later :trollface:

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/EnvironmentInterface.scala (hunk 2 above)
Comment (cocreature) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/EnvironmentInterface.scala:29]: why? Doing it in one go seems simple enough and should be faster

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/Interface.scala (hunk 1 above)
Comment (cocreature) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/Interface.scala:45]: everything gets better if you add another type parameter.

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/EnvironmentInterface.scala (hunk 2 above)
Comment (cocreature) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/EnvironmentInterface.scala]: oh didn’t realize updating maps was so much more expensive. changed 9dd1a3f91d29a03567eec4f4746632a70d37faf6

Comment (S11001001) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/EnvironmentInterface.scala]: I'm not generally opposed to updating maps iteratively, if it's clearer, but when I do, I try to use `scalaz.==>>` instead, which is guaranteed to behave like `Data.Map.Strict`.

**Review by S11001001 — APPROVED** [review]

**Pull request merged**